File: src/net/merayen/elastic/ui/objects/top/views/filebrowserview/FileList.java
package net.merayen.elastic.ui.objects.top.views.filebrowserview;

import java.io.File;
import java.util.ArrayList;

import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.objects.components.Scroll;
import net.merayen.elastic.ui.objects.components.autolayout.AutoLayout;
import net.merayen.elastic.ui.objects.components.autolayout.LayoutMethods;

class FileList extends UIObject {
	interface Handler {
		public void onSelect(File file);
	}

	float width = 100;
	float height = 100;

	private class FList extends AutoLayout {
		private File[] current_files;

		FList() {
			super(new LayoutMethods.HorizontalBox(10));
		}

		@Override
		public void onUpdate() {
			((LayoutMethods.HorizontalBox) placement).setMaxWidth(width);

			if(files != current_files) {
				for(UIObject obj : new ArrayList<>(getSearch().getChildren()))
					remove(obj);

				add(new FileListItem() {{
					label = ".. (go back)";
					setHandler(() -> {
						if(handler != null)
							handler.onSelect(null);
					});
				}});

				if(files != null) {
					for(File f : files) {
						add(new FileListItem() {{
							label = f.getName();
							setHandler(() -> {
								if(handler != null)
									handler.onSelect(f);
							});
						}});
					}
				}

				current_files = files;
			}

			super.onUpdate();
		}
	}

	private final FList flist = new FList();
	private final Scroll scroll = new Scroll(flist);
	private File[] files;
	private String path;
	private Handler handler;

	@Override
	public void onInit() {
		add(scroll);
	}

	@Override
	public void onUpdate() {
		scroll.setLayoutHeight(height);
		scroll.setLayoutWidth(width);
	}

	void browse(String path) {
		this.path = path;
		files = new File(path).listFiles();
	}

	void setHandler(Handler handler) {
		this.handler = handler;
	}

	void up() {
		File parent = new File(path).getParentFile();
		if(parent != null)
			browse(parent.getAbsolutePath());
	}
}

File: src/net/merayen/elastic/ui/objects/top/views/filebrowserview/FileList.kt
package net.merayen.elastic.ui.objects.top.views.filebrowserview

import java.io.File
import java.util.ArrayList

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.objects.components.Scroll
import net.merayen.elastic.ui.objects.components.autolayout.AutoLayout
import net.merayen.elastic.ui.objects.components.autolayout.LayoutMethods

internal class FileList : UIObject() {

	var layoutWidth = 100f
	var layoutHeight = 100f

	private val flist = FList()
	private val scroll = Scroll(flist)
	var files: Array<File>? = null
	private var path: String? = null
	private var handler: Handler? = null

	internal inner class FList internal constructor() : AutoLayout<LayoutMethods.HorizontalBox>(LayoutMethods.HorizontalBox(10f)) {
		private var current_files: Array<File>? = null

		override fun onUpdate() {
			(placement as LayoutMethods.HorizontalBox).maxWidth = getWidth()

			if (files != current_files) {
				for (obj in ArrayList(search.children))
					remove(obj)

				add(object : FileListItem(File(".. (go back)"), false) {
					init {
						setHandler(object : FileListItem.Handler {
							override fun onClick() {
								if (handler != null)
									handler!!.onSelect(null)
							}
						})
					}
				})

				if (files != null) {
					for (f in files!!) {
						add(object : FileListItem(f, true) {
							init {
								setHandler(object : FileListItem.Handler {
									override fun onClick() {
										if (handler != null)
											handler!!.onSelect(f)
									}
								})
							}
						})
					}
				}

				current_files = files
			}

			super.onUpdate()
		}
	}

	internal interface Handler {
		fun onSelect(file: File?)
	}

	override fun onInit() {
		add(scroll)
	}

	override fun onUpdate() {
		scroll.layoutWidth = layoutWidth
		scroll.layoutHeight = layoutHeight
	}

	override fun onDraw(draw: Draw) {
		draw.setColor(255, 0, 255)
		draw.rect(0f, 0f, layoutWidth, layoutHeight)
	}

	fun browse(path: String) {
		this.path = path
		files = File(path).listFiles()
	}

	fun setHandler(handler: Handler) {
		this.handler = handler
	}

	fun up() {
		val parent = File(path!!).parentFile
		if (parent != null)
			browse(parent!!.absolutePath)
	}
}
